Add documentation for hoomd.Operations

b-butler opened the pull request:
## Description

Add documentation for `hoomd.Operations` class and `hoomd.operations` module.
I also fix a bug where `simulation.operations += operation` would raise an error.

## Motivation and context

The class needed documentation.

## How has this been tested?

Passes existing tests; the file passes `pydocstyle` checking; and sphinx builds without errors.

## Checklist:

- [x] I have reviewed the [**Contributor Guidelines**](https://github.com/glotzerlab/hoomd-blue/blob/master/CONTRIBUTING.md).
- [x] I agree with the terms of the [**HOOMD-blue Contributor Agreement**](https://github.com/glotzerlab/hoomd-blue/blob/master/ContributorAgreement.md).
- [ ] My name is on the list of contributors (`sphinx-doc/credits.rst`) in the pull request source branch.

bdice commented on hoomd/operations.py:
```diff
@@ -117,20 +197,50 @@ def integrator(self, op):
+        """
         if isinstance(operation, hoomd.integrate._BaseIntegrator):
             raise ValueError(
                 "Cannot remove iterator without setting to a new integator.")
```
This method needs to raise a `TypeError` after the `elif`s if the argument is not an operation class!

b-butler replied:
Do we need to error here? It for sure makes sense to error in `add`, but here I think it makes sense that nothing happens if an user calls `operations.remove(3)` since `3 not in operations`.

bdice replied:
Trying to remove an object that doesn't exist in the collection typically raises an exception. In the case where none of the `isinstance` checks pass, I would expect a `TypeError` instead of a `ValueError`. However, I would expect a `ValueError` if one tries to remove an updater/tuner/analyzer that isn't actually in this `Operations` container.
```pytb
>>> data = [1, 2, 3]
>>> data.remove('potato')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: list.remove(x): x not in list
```

joaander replied:
Why can't you remove an integrator without setting a new integrator? The default integrator is None in Python (a NULL shared pointer in C++).

b-butler replied:
We could do that then. I didn't know we were accepting that as a valid state.

joaander replied:
For no other reason than it is a valid default state, so we should allow users to return the object to its default state after modifying it.

b-butler replied:
I agree and have modified it accordingly. I meant I didn't know it was a valid C++ state.

bdice commented on hoomd/operations.py:
```diff
@@ -142,4 +252,11 @@ def remove(self, operation):
+        """Works the same as `Operations.remove`.
+
+        Args:
+            operation (``operation``): A HOOMD-blue updater, analyzers, compute,
```
```suggestion
            operation (``operation``): A HOOMD-blue updater, analyzer, compute,
```

bdice commented on hoomd/simulation.py:
```diff
@@ -115,6 +116,30 @@ def state(self):
+                self._operations = operations
+
     def sanity_check(self):
         raise NotImplementedError
```
This seems like an oversight. This method should probably be private, or removed?

b-butler replied:
It should be removed, but this is not really related to the PR at hand. There are multiple other methods that likely should be removed from `hoomd.simulation.Simulation`.

joaander replied:
#783 cleans removes these methods from `Simulation`

bdice reviewed (changes requested):
See comments below.

b-butler commented on hoomd/operations.py:
```diff
@@ -1,3 +1,10 @@
+
+Defines the `Operations` class which serves as the main class for storing and
+organizing the many parts of a simulation in a way that allows parts to be added
+and removed before and after _attaching_ C++ objects.
```
We don't export the module docstring for `operations.py`, and t is an implementation detail. I put it there more for developers. I should just make a comment at the top of the file in case anyone uses this module.

b-butler commented on hoomd/operations.py:
```diff
@@ -7,12 +14,40 @@
     return (value._cpp_obj, value.trigger)
 
 
 class Operations:
```
That is not a bad idea. We could make it a `Collection`. We don't get any mixin methods for that though. On the `MutableSet` set idea, it is in a lot of ways like a set, but then using `+=` as a valid operator for the object may feel off for users since Python sets don't implement a `__iadd__`. Also, `-=` would change meaning and only work with other `Operations` objects if we are being true to the set concept.

bdice replied:
I vote to make it a `Collection` but not a set or anything more specific.

b-butler commented on hoomd/operations.py:
```diff
@@ -7,12 +14,40 @@
+    see examples of these types of operations see (`hoomd.update`, `hoomd.tune`,
+    `hoomd.dump`, and `hoomd.md.thermo`).
+
+    Operations can only be added once to an `Operations` object. This includes
+    the same object. That means to have multiple of the same operation in an
+    `Operations` object, multiple instances of the operation class would need to
+    be instantiated.
```
1. yes
2. yes
What cannot happen is the same instance being added twice to the same or different `Operations` instances.

b-butler commented on hoomd/operations.py:
```diff
@@ -7,12 +14,40 @@
+    `Operations` object, multiple instances of the operation class would need to
+    be instantiated.
+
+    Note:
+        An `Operations` object is created by default when a new simulation is
+        created.
```
Yes they can.

b-butler commented on hoomd/operations.py:
```diff
@@ -24,21 +59,36 @@ def __init__(self, simulation=None):
 
         self._tuners.append(ParticleSorter())
```
They can, and it isn't dangerous, but it would significantly slow down their simulation. We expose the object because there are properties that users can usefully modify.

bdice replied:
I would expect that instantiating a new `Operations` instance is the same as calling `clear` on all the lists of an existing one. That is, I expect both objects to be empty. However, clearing the list of tuners will also delete the particle sorter. 😕 I recommend one or more of the following:
1. Document clearly that newly-instantiated `Operations` objects have a `ParticleSorter` in the tuners by default.
2. Ensure that a `ParticleSorter` is automatically created/added at runtime if one does not exist (or at least warn that a `ParticleSorter` is not present). This assumes that users always want one, which I think is true?

joaander replied:
We should document it explicitly.

Users always want to use the sorter, it always improves performance, and they should be able to change the sorter parameters before scheduling. The unit tests use this functionality to reduce the grid size and improve startup time. Users may want to tweak grid parameters or the period to potentially impact performance.

b-butler replied:
I am not sure we want to do 2 except to warn users. If they explicitly remove the particle sorter, we should respect that.

b-butler commented on hoomd/operations.py:
```diff
@@ -51,6 +101,21 @@ def _sys_init(self):
+
+        This is provided in the public-API to allow users to get quantities that
+        otherwise would not be available until after a call to
+        `hoomd.Simulation.run`. This does not have to be called before a ``run``
+        call as this will be called automatically if necessary.
```
Conditionally, there is a `scheduled` property that tells us the state of the instance.

b-butler commented on hoomd/operations.py:
```diff
@@ -51,6 +101,21 @@ def _sys_init(self):
+
+        The function creates many C++ objects in the background to prepare for
+        running a simulation. Before this all operations exist purely as Python
+        objects.
```
It is more to try and explain why anything needs to happen at all before a run call or why `lj.energies` is `None` before scheduling. I wasn't sure how to explain that without going into some implementation details.

b-butler commented on hoomd/operations.py:
```diff
@@ -117,20 +197,50 @@ def integrator(self, op):
 
     @property
     def updaters(self):
+        """list[``Updater``]: A list of updater operations.
```
The actual base class is private, so I was avoiding linking to an actual class.

And yes.

b-butler replied:
We use a lot of classes that users should never instantiate, and we normally make them private to signify that. However, they have to be exposed to the user for them to use. We could make the classes public, but then we would need to warn that they should never be instantiated.

b-butler replied:
I do like the `List[_Updater]` it matches Python's own typing system, but sphinx does not know how to properly link to the standard Python documentation. Though, this is a decision we should discuss.

The reason I choose not to document the actual type here is that it appears as an ordinary list to users (besides the type of course). The other private classes we document are to prevent redocumenting a large number of methods which isn't the case here.  We could document `_Updaters` and the like regardless.

b-butler commented on hoomd/operations.py:
```diff
@@ -117,20 +197,50 @@ def integrator(self, op):
     def remove(self, operation):
+        """Remove an operation from the `Operations` object.
+
+        Checks for removal according to object id.
```
This is more if they create two identical operations and use one to try to remove the other it won't work. I wanted to document that is expected behavior.

joaander replied:
Unfortunately, this is needed. Python list objects remove the first object that is equal to the requested object and we define `__eq__` for operations to be True when their parameters are identical.

HOOMD users are Python programmers and should be assumed to understand what an `id` is. You could reword to match the language used in Python (https://docs.python.org/3/tutorial/datastructures.html):
```
Remove the item from the collection whose id is the same as *operation*.
```

b-butler commented on hoomd/operations.py:
```diff
@@ -117,20 +197,50 @@ def integrator(self, op):
+        """Works the same as `Operations.add`.
+
+        Args:
+            operation (``operation``): A HOOMD-blue updater, analyzers, compute,
```
Not one that is public.

b-butler replied:
We could do that. I want to keep the private implementation side of the API as far from the users as possible, but that is not reasonably doable in some cases. Some of our *private* classes are essentially public with the caveat that users should not attempt to instantiate them.

bdice replied:
I think it's reasonable to document the private base type here, rather than a placeholder type that can't be found in the source code at all.

joaander commented on hoomd/operations.py:
```diff
@@ -76,9 +142,16 @@ def _store_reader(self, reader):
         pass
 
     def __contains__(self, obj):
+        """Whether an operation is contained in the object.
```
`Operations` is more like a sequence than a set. `updaters`, `analyzers`, and `tuners` are lists as they have a well defined order. `Simulation` evaluates the these operations in a well defined order as documented in #783:
```
            while timestep < end_step:
                for tuner in operations.tuners:
                    if tuner.trigger(timestep):
                        tuner.tune(timestep)
                for updater in operations.updaters:
                    if updater.trigger(timestep):
                        updater.update(timestep)
                if operations.integrator is not None:
                    operations.integrator(timestep)
                timestep += 1
                for writer in operations.writers:
                    if writer.trigger(timestep):
                        writer.write(timestep)
```
Given this, I would interpret the operations list as concatenation of the `tuners`, `updaters`, [integrator], and `writers` (`analyzers` currently as this has not yet been renamed.)

b-butler replied:
We should reorder `__iter__` and the docs according to #783.

joaander commented on hoomd/operations.py:
```diff
@@ -7,12 +14,40 @@
+    The types of operations which can be added to an `Operations` object are
+    integrators, updaters, analyzers, tuners, and computes. An `Operations` can
+    only every hold one integrator at a time. On the other hand, an `Operations`
+    object can hold any number of updaters, analyzers, tuners, or computes. To
```
The order is defined by the order in which these operations are called during the run loop, documented in #783. Also, the introduction to HOOMD-tutorial introduces the concepts of updaters, analyzers, and the integrator as core concepts - so new HOOMD users should be familiar with these: https://hoomd-blue.readthedocs.io/en/feature-new-object-api/tutorial/00-Introducing-HOOMD-blue/01-The-Simulation-Object.html and later sections.

Note that all of the convenience methods that make `Operations` itself appear like a sequence or collection introduce a number of fluid representations so I avoid them in the tutorial and always use the explicit e.g. `sim.operations.tuners.append(tuner)`. The individual `tuners`, `updaters`, etc.. objects are essentially lists which the new user is expected to be familiar with.

I see all these `Operations` convenience methods more as a shortcut for experienced users.

b-butler replied:
I agree they are more conveniences that make some manipulations easier.

joaander commented on hoomd/operations.py:
```diff
@@ -75,29 +150,43 @@ def _store_reader(self, reader):
+        """Iterates through all contained operations."""
+        integrator = (self._integrator) if self._integrator else []
+        yield from chain(
+            integrator, self._analyzers, self._updaters, self._tuners)
```
In #574, we should adjust this order to match the order in which the operations will be evaluated during the run loop.

joaander reviewed (changes requested):
Thanks! I have a few changes noted below:

b-butler commented on hoomd/operations.py:
```diff
@@ -7,12 +14,40 @@
+
+    The types of operations which can be added to an `Operations` object are
+    integrators, updaters, analyzers, tuners, and computes. An `Operations` can
+    only every hold one integrator at a time. On the other hand, an `Operations`
```
It can hold ``None``.

b-butler commented on hoomd/operations.py:
```diff
@@ -76,9 +142,16 @@ def _store_reader(self, reader):
+
+        Args:
+            obj (``any``): Returns whether this exact operation is contained in
+                the object.
```
The reason was that an equivalent operation won't return `True`.

b-butler commented on hoomd/operations.py:
```diff
@@ -75,29 +150,43 @@ def _store_reader(self, reader):
+
+        `Operations` objects have an initial ``integrator`` property of
+        ``None``. Can be set to MD or HPMC integrators. The property can also be
+        set to ``None``, though this will raise errors when scheduled.
```
Okay I will change the method accordingly.

bdice commented on hoomd/operations.py:
```diff
@@ -1,47 +1,112 @@
+        `Operations` instance.
+
+        Args:
+            operation (``operation``): A HOOMD-blue updater, analyzers, compute,
```
```suggestion
            operation (``operation``): A HOOMD-blue updater, analyzer, compute,
```

bdice commented on hoomd/operations.py:
```diff
@@ -1,47 +1,112 @@
-            raise ValueError("Operation is not of the correct type to add to"
-                             " Operations.")
+            raise TypeError(
+                "Operation is not of the correct type to add to Operations.")
```
I find that printing invalid types in messages like these helps users debug their code much more efficiently. It helps users reason about what went wrong.
```suggestion
                f"Type {type(operation)} is not a valid type to add to Operations.")
```

bdice commented on hoomd/operations.py:
```diff
@@ -51,6 +116,17 @@ def _sys_init(self):
     def schedule(self):
+        """Prepares all operations for a `hoomd.Simulation.run` call.
+
+        This is provided in the public API to allow users to access quantities
```
If it's possible to automatically call this before running, why isn't it automatically called when attempting to access a quantity that requires scheduling? Then this function wouldn't need to be in the public API at all, and it's entirely transparent to users. (There may be issues I'm not aware of with this suggestion, since I don't know what requires scheduling.)

joaander replied:
There are many quantities (i.e. the `kinetic_energy` property computed by `ThermoDynamicQuantities`) that are undefined when the operation is not scheduled (they return `None`). In the rare case that a user wants to compute these values before a call to `run`, they need to call `schedule` explicitly first. 

If these properties instead implicitly called `schedule`, one of the unintended side effects would be the resolution of all `TypeParameters` down C++. 1) This may lead to errors or warnings e.g. in pair.LJ parameters. 2) The types in the parameters are reduced to only those in the system. I think it would be awkward and confusing for users if a call to `thermo.kinetic_energy` resulted in errors from `pair.LJ`. For point 2, this potentially prevents use-cases like we employ in MoSDeF where we are using the `Logger` infrastructure to write out force field metadata, or write_debug_data. The output would depend on the order of items in the logger - e.g. if pair.LJ type parameters were written first they would have all types specified in the force field, but if the system potential energy is written first then the parameters will be collapsed.

Given the wide range of possible side effects, I think it best to make scheduling explicit for these cases and only implicit in the common case of `run`. Methods that require scheduling will be documented and raise an appropriate exception and we can clearly document all properties that return `None` before scheduling.

joaander replied:
Alternately, we could make schedule private and document these as being available after run. Users can call `sim.run(0)` to achieve this without advancing the system forward. `sim.run(0)` performs some extra steps like communication and integrator preparation that are not performed by `schedule` - see e.g. #726.

bdice replied:
@joaander I like your idea of `sim.run(0)`. I think that is conceptually easier to understand than the concept of "scheduling" (and explaining when it is or isn't necessary). I believe it is also closer to the current model of HOOMD 2?

bdice commented on hoomd/operations.py:
```diff
@@ -65,6 +141,7 @@ def schedule(self):
         self._scheduled = True
 
     def unschedule(self):
+        """Undo the effects of `Operations.schedule`."""
```
When would a user want to use this method? A little more documentation would be helpful.

b-butler replied:
A user in general should not have to deal with this especially if we use `sim.run(0)`.

bdice commented on hoomd/operations.py:
```diff
@@ -117,29 +213,76 @@ def integrator(self, op):
 
     @property
     def analyzers(self):
+        """list[`hoomd.operation.Analzyer`]: A list of analyzer operations.
```
```suggestion
        """list[`hoomd.operation.Analyzer`]: A list of analyzer operations.
```

bdice commented on hoomd/operations.py:
```diff
@@ -75,38 +152,57 @@ def _store_reader(self, reader):
 
     @property
     def integrator(self):
+        """``Integrator``: An MD or HPMC integrator object.
```
Is this formatted correctly? The other Sphinx class references use one backtick, I believe. Unless this is supposed to be a reference to a private base class like `_BaseIntegrator` (see comment below about documenting private base types, it seems that @joaander and I agree this is a good idea).

joaander replied:
Yes, we discussed this on Monday. We will document these core base classes project-wide in a separate PR.

bdice commented on hoomd/operations.py:
```diff
@@ -117,29 +213,76 @@ def integrator(self, op):
+                analyzer, tuner, or compute, to remove from the container.
+
+        Raises:
+            ValueError: raises if operation is not found in this container.
+            TypeError: raises if operation is not of a valid type for this
+                container.
```
I'm adopting the "Raises" phrasing structure from [numpydoc](https://numpydoc.readthedocs.io/en/latest/format.html) in this suggestion. I prefer its conciseness.
```suggestion
            ValueError: If ``operation`` is not found in this container.
            TypeError: If ``operation`` is not of a valid type.
```

bdice commented on hoomd/operations.py:
```diff
@@ -1,47 +1,112 @@
+                tuner, or integrator to add to the collection.
+
+        Raises:
+            RuntimeError: raised when an operation belonging to to this or
+                another `Operations` instance is passed.
+            TypeError: raised when the passed operation is not of a valid type.
```
I'm adopting the "Raises" phrasing structure from [numpydoc](https://numpydoc.readthedocs.io/en/latest/format.html) in this suggestion. I prefer its conciseness. Also I think the `RuntimeError` should be a `ValueError` instead. It's an invalid value. Examples of `RuntimeError`s I can come up with seem more obscure and are usually not related to adding/setting a value.
```suggestion
            ValueError: If ``operation`` already belongs to this or another `Operations` instance.
            TypeError: If ``operation`` is not of a valid type.
```

bdice commented on hoomd/operations.py:
```diff
@@ -1,47 +1,112 @@
         # we raise this error here to provide a more clear error message.
-        if op._added:
+        if operation._added:
             raise RuntimeError(
```
```suggestion
            raise ValueError(
```

bdice commented on hoomd/operations.py:
```diff
@@ -1,47 +1,112 @@
-        if op._added:
+        if operation._added:
             raise RuntimeError(
                 "Operation cannot be added to twice to operation lists.")
```
The previous text doesn't make sense and could be more specific.
```suggestion
                "The provided operation has already been added to an Operations instance.")
```

bdice commented on hoomd/operations.py:
```diff
@@ -117,29 +213,76 @@ def integrator(self, op):
         self.add(operation)
+        return self
 
     def remove(self, operation):
```
I'm confused by the long distance between "add" and "remove." Shouldn't these methods be near each other?

bdice commented on hoomd/operations.py:
```diff
@@ -117,29 +213,76 @@ def integrator(self, op):
             self._tuners.remove(operation)
+        else:
+            raise TypeError(
+                "operation is not a valid type for an Operations container.")
```
```suggestion
                f"Type {type(operation)} is not a valid type to remove from Operations.")
```

bdice commented on hoomd/operations.py:
```diff
@@ -117,29 +213,76 @@ def integrator(self, op):
-                "Cannot remove iterator without setting to a new integator.")
-        elif isinstance(operation, _Analyzer):
+            self.integrator = None
+        elif isinstance(operation, Analyzer):
```
I haven't looked to see how many times this long `elif` pattern is repeated, but it might be worth having some kind of dispatch mechanism:
```python
# Integrators are probably treated differently...
if isinstance(operation, hoomd.integrate._BaseIntegrator):
    # stuff
# ...but all the other operations have a lot of copy-paste that could be eliminated
elif isinstance(operation, _Operation):
    container = self._container_for_operation_type(operation)
    container.remove(operation)

...

def _container_for_operation_type(self, operation):
    if isinstance(operation, Analyzer):
        return self._analyzers
    elif ...
```

bdice reviewed (changes requested):
Good progress overall. I have a second round of comments here. I think this is probably my last round of feedback on this PR. If you want a final round of review, please resolve (or respond to) all open comments before re-requesting. There were a lot of "outdated" comments left on the Conversation page from the first review.

joaander commented on hoomd/operation.py:
```diff
@@ -515,12 +515,39 @@ def _attach(self):
+    """Base class for all HOOMD tuners.
+
+    A tuner is an operation which tunes the parameters of another operation for
+    performance or other reasons. A tuner does not properly modify the
+    simulation state (e.g. move particles or change bonds).
```
What Brandon is trying to account for here is that `ParticleSorter` is a `Tuner` that modifies the order of the particles in the `State`, however this does not modify the microstate of the system. The problem is a fluid representation with the word "State" meaning two different things. We could reword to explicitly allow this: "A `Tuner` does not modify the simulation `State`, except to change the ordering of elements in local snapshots."

b-butler commented:
@bdice I requested your review since you had requested changes. If you don't want to re-review (I have implemented your proposed changes or addressed your comments) feel free to remove yourself.

b-butler commented:
@joaander this is ready for another round of review.

joaander commented on hoomd/operations.py:
```diff
@@ -1,49 +1,177 @@
+    only ever hold one integrator at a time. On the other hand, an `Operations`
+    object can hold any number of tuners, updaters, analyzers, or computes. To
+    see examples of these types of operations see `hoomd.tune` (tuners),
+    `hoomd.update` (updaters), `hoomd.hpmc.integrate` or `hoomd.md.integrate`
```
```suggestion
    `hoomd.update` (updaters), `hoomd.hpmc.integrate` or `hoomd.md.Integrator`
```

joaander commented on hoomd/operations.py:
```diff
@@ -1,49 +1,177 @@
+    same `Operations` container more than once.
+
+    All `Operations` instances start with a `hoomd.tune.ParticleSorter` instance
+    in their ``tuners`` attribute. This is vital for increasing simulation
```
```suggestion
    in their ``tuners`` attribute. This increases simulation
```

joaander commented on hoomd/operations.py:
```diff
@@ -1,49 +1,177 @@
+                f"{type(operation)} is not a valid operation type.")
+
+    def add(self, operation):
+        """Add operation to this container.
```
```suggestion
        """Add an operation to this container.
```

joaander commented on hoomd/operations.py:
```diff
@@ -1,49 +1,177 @@
+    def add(self, operation):
+        """Add operation to this container.
+
+        Adds the provide operation to the appropriate attribute of the
```
```suggestion
        Adds the provided operation to the appropriate attribute of the
```

joaander commented on sphinx-doc/module-hoomd-operation.rst:
```diff
@@ -0,0 +1,19 @@
+.. autosummary::
+    :nosignatures:
+
+    Analyzer
```
We should also document `Operation` and `Compute`. Were you planning this in another PR?

b-butler replied:
Operation yes, compute no. I will add it now.

joaander reviewed (approved):
Thanks! This looks great. I have a few small suggestions to improve.

b-butler commented:
@bdice has opted out of reviewing again so we should ignore his request changes.
